### PR TITLE
Backwardscompatibly handle syscalls unknown to us or libseccomp

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/shared/seccomp-util.c [hunk 1]
```diff
@@ -1129,7 +1129,9 @@ int seccomp_load_syscall_filter_set_raw(uint32_t default_action, Hashmap* filter
 
                 log_trace("Operating on architecture: %s", seccomp_arch_to_string(arch));
 
-                r = seccomp_init_for_arch(&seccomp, arch, default_action);
+                /* We install ENOSYS as the default action, but it will only apply to syscalls which are not
+                 * in the @known set. */
+                r = seccomp_init_for_arch(&seccomp, arch, SCMP_ACT_ERRNO(ENOSYS));
```
Comment (YHNdnzj) [inline src/shared/seccomp-util.c:1134]: With this, I guess we can remove the `#else` branch for `#if defined(__SNR_fchmodat2)`?

**Review by bluca — APPROVED** [review]

File: src/shared/seccomp-util.c [hunk 2]
```diff
@@ -1164,6 +1166,23 @@ int seccomp_load_syscall_filter_set_raw(uint32_t default_action, Hashmap* filter
...
                 }
 
+                NULSTR_FOREACH(name, syscall_filter_sets[SYSCALL_FILTER_SET_KNOWN].value) {
+                        int id;
+
+                        id = seccomp_syscall_resolve_name(name);
+                        if (id < 0)
+                                continue;
+
+                        /* Ignore the syscall is it was already handled above */
```
Comment (YHNdnzj) [inline src/shared/seccomp-util.c]:
```suggestion
                        /* Ignore the syscall if it was already handled above */
```

**Review by YHNdnzj — APPROVED** [review]

File: src/shared/seccomp-util.c [hunk 3]
```diff
@@ -1164,6 +1166,23 @@ int seccomp_load_syscall_filter_set_raw(uint32_t default_action, Hashmap* filter
...
+                        if (hashmap_contains(filter, INT_TO_PTR(id + 1)))
+                                continue;
+
+                        r = seccomp_rule_add_exact(seccomp, default_action, id, 0);
+                        if (r < 0 && r != -EDOM)  /* EDOM means that the syscall is not available for arch */
+                                return log_debug_errno(r, "Failed to add rule for system call %s() / %d: %m",
+                                                       name, id);
```
Comment (poettering) [inline src/shared/seccomp-util.c:1183]: so in the nspawn case you also added this:

d75615f398a0fbf986cf03924462863ca6ee2f9f

we might want to do the same here? (not sure why this isn't the default in libseccomp anyway? seems so pointless having to enable this)

File: src/shared/seccomp-util.c (hunk 1 above)
Comment (keszybz) [reply src/shared/seccomp-util.c:1134]: Indeed.

**Comment by keszybz** [comment]
Updated with the suggested changes.

**Comment by keszybz** [comment]
Ooops. The change that was suggested in https://github.com/systemd/systemd/pull/30291#pullrequestreview-1760421195 broke a valid test case. I removed that that chunk and added a note in the commit message. In this regard, it's a revert to the previous version of the patch.

**Pull request merged**